fvm/src/syscalls/mod.rs

```diff
@@ -91,42 +89,62 @@ pub fn charge_for_exec<K: Kernel>(
 
     let data = ctx.data_mut();
 
-    // Separate the amount of gas charged for memory and apply discount on first page.
-    let memory_price = data
-        .kernel
-        .price_list()
-        .wasm_rules
-        .memory_expansion_per_byte_cost;
-
+    // Separate the amount of gas charged for memory; this is only makes a difference in tracing.
     let memory_bytes = data.kernel.limiter_mut().total_exec_memory_bytes();
-    let memory_delta_bytes = memory_bytes - data.last_memory_bytes;
-    let mut memory_gas = memory_price * memory_delta_bytes as i64;
-    exec_gas = (exec_gas - memory_gas).max(Gas::zero());
-
-    // The memory grows by at least one page, so if we're haven't yet charged for any memory,
-    // then this is the time to apply the first-page discount.
-    if data.last_memory_bytes == data.start_memory_bytes && memory_bytes > data.start_memory_bytes {
-        let free_memory_gas = memory_price * wasmtime_environ::WASM_PAGE_SIZE as i64;
-        memory_gas = (memory_gas - free_memory_gas).max(Gas::zero());
-    }
+    let memory_delta_bytes = (memory_bytes - data.last_memory_bytes).max(0);
+    let memory_gas = data.kernel.price_list().grow_memory_gas(memory_delta_bytes);
 
-    data.last_memory_bytes = memory_bytes;
-
-    // The separation of wasm_exec and wasm_memory is optional, it might be nice to have for statistics.
+    exec_gas = (exec_gas - memory_gas).max(Gas::zero());
 
     data.kernel
         .charge_gas("wasm_exec", exec_gas)
         .map_err(Abort::from_error_as_fatal)?;
 
     if !memory_gas.is_zero() {
         data.kernel
-            .charge_gas("wasm_memory", memory_gas)
+            .charge_gas("wasm_memory_grow", memory_gas)
             .map_err(Abort::from_error_as_fatal)?;
     }
+    data.last_memory_bytes = memory_bytes;
 
     Ok(())
 }
 
+/// Charge for the initial memory before a Wasm module is instantiated.
+///
+/// The Wasm instrumentation machinery via [fvm_wasm_instrument::gas_metering::MemoryGrowCost]
+/// only charges for growing the memory _beyond_ the initial amount. It's up to us to make sure
+/// the minimum memory is properly charged for.
+pub fn charge_for_init<K: Kernel>(
+    ctx: &mut impl AsContextMut<Data = InvocationData<K>>,
+    module: &Module,
+) -> crate::kernel::Result<()> {
+    let min_memory_bytes = min_memory_bytes(module)?;
+    let mut ctx = ctx.as_context_mut();
+    let kernel = &mut ctx.data_mut().kernel;
+    let charge = kernel.price_list().init_memory_gas(min_memory_bytes);
+    kernel.charge_gas("wasm_memory_init", charge)?;
+
+    // Adjust `last_memory_bytes` so that we don't charge for it again in `charge_for_exec`.
+    ctx.data_mut().last_memory_bytes += min_memory_bytes;
```

Comment:
Switching over to instrumentation will let us get rid of this.

Reply:
I still don't understand. We _are_ using instrumentation because of [this rule](https://github.com/filecoin-project/ref-fvm/blob/d00a3ea4cebbd50e579d613f03deb0b7098202e9/fvm/src/gas/price_list.rs#L743-L750) triggering the [injection](https://github.com/filecoin-project/fvm-wasm-instrument/blob/master/src/gas_metering/mod.rs#L200) of the grow cost function. 

But my understanding is that this gas counting happens purely inside Wasm, and it's the `charge_for_exec` method which [reads the current value and compares to the last](https://github.com/filecoin-project/ref-fvm/blob/d00a3ea4cebbd50e579d613f03deb0b7098202e9/fvm/src/syscalls/mod.rs#L76-L88) to figure out what it actually has to [charge for](https://github.com/filecoin-project/ref-fvm/blob/d00a3ea4cebbd50e579d613f03deb0b7098202e9/fvm/src/syscalls/mod.rs#L100). 

Charging for memory is using the same accounting mechanism, so we have to charge for it the same way. 

We can get rid of `last_memory_bytes` even now if we don't want to have a separate `wasm_exec` and `wasm_memory_grow` item in the tracer, because that is the only reason it exists.

Reply:
Just let me know if you thing that separation is not needed, I only put it in there to get feedback about it in the PR, present it as an option.